README.md

```diff
@@ -4,15 +4,39 @@
+### Functionality and Extension Capabilities
+Each time a balance of a connected account changes, the token contract notifies all plugins that the account has subscribed to. Each plugin implements the `_updateBalances` function, which processes change requests for a balance alteration, inclusive of balance change data (from/to addresses and amount). The plugin can then perform necessary actions according to its logic to extend the base token functionality.
 
-Contribution:
-- Install dependencies: `yarn` 
-- Run tests: `yarn test` or `run test:ci`
+Keep in mind that the plugin's processing logic consumes additional gas. Although there's a limit on the amount of gas that a plugin can expend, the account owner has autonomy in selecting which plugins to subscribe their account to.
+
+### Examples of Plugins
+* [FarmingPlugin](https://github.com/1inch/farming) - *The 1inch staking reward farming is based on this plugin.*
+This plugin facilitates farming without the need for asset transfer or locking on a farm. An account owner can join the farm by simply adding this plugin to their account. The plugin handles all necessary accounting for joined farmers and distributes rewards based on the actual farming token balance of the account during the farming period. This is particularly useful for non-transferable tokens, such as debt tokens in lending protocols.
+
+* [DelegatingPlugin](https://github.com/1inch/delegating) - *The 1inch fusion mode resolver delegation is based on this plugin.*
+ This plugin allows an account balance to be delegated to another address. This can be advantageous for governance tokens if an account owner wants to delegate their voting power to another account without a physical token transfer. The owner can recall or redelegate their delegation at any time without locking their tokens in a governance contract.
+
+ ### Security Measures
+The plugin system operates under the assumption that an account may subscribe to a malicious plugin. Consequently, the following restrictions apply to each call:
+
+* If a plugin fails its execution and reverts, it won't impact the main flow. The failed plugin is bypassed, and execution continues.
+* Each plugin has a gas limit set by the parent contract. If there is insufficient gas, it won't affect the main flow's execution.
+* Plugins are executed in the context specified in their contract (parent contract uses a call, not delegatecall).
+* Plugins cannot alter the calling contract state
+* Plugins cannot be reentered
+
+### Integrating Plugin Support in Your Token Implementation
+
+1. Inherit token using `contract MyToken is ERC20Plugins { ... }`.
+Or wrap it using `contract MyWrapper is ERC20Wrapper, ERC20Plugins { ... }`.
+This adds support for the plugin infrastructure.
+2. Create and deploy a plugin smart contract or choose an existing deployed plugin to add to wallets holding your token, then add the plugin to a wallet using `MyToken.addPlugin(plugin)`.
```

Comment:
`MyToken.addPlugin(plugin)` should be called by each user who wants to add it, right? I would then move the part about plugin adding to a separate point, changing the context a bit: "User should add the plugin to their wallet calling `MyToken.addPlugin(plugin)`" or something.